src/EditorFeatures/Core/Implementation/CommentSelection/ToggleLineCommentCommandHandler.cs

```diff
@@ -0,0 +1,191 @@
+
+        public override string DisplayName => EditorFeaturesResources.Toggle_Line_Comment;
+
+        protected override string GetTitle(ValueTuple command) => EditorFeaturesResources.Toggle_Line_Comment;
+
+        protected override string GetMessage(ValueTuple command) => EditorFeaturesResources.Toggling_line_comment;
+
+        internal async override Task<CommentSelectionResult> CollectEditsAsync(Document document, ICommentSelectionService service,
+            ITextBuffer subjectBuffer, NormalizedSnapshotSpanCollection selectedSpans, ValueTuple command, CancellationToken cancellationToken)
+        {
+            var experimentationService = document.Project.Solution.Workspace.Services.GetRequiredService<IExperimentationService>();
+            if (!experimentationService.IsExperimentEnabled(WellKnownExperimentNames.RoslynToggleLineComment))
+            {
+                return s_emptyCommentSelectionResult;
+            }
+
+            var commentInfo = await service.GetInfoAsync(document, selectedSpans.First().Span.ToTextSpan(), cancellationToken).ConfigureAwait(false);
+            if (commentInfo.SupportsSingleLineComment)
+            {
+                return ToggleLineComment(commentInfo, selectedSpans);
+            }
+
+            return s_emptyCommentSelectionResult;
+        }
+
+        private CommentSelectionResult ToggleLineComment(CommentSelectionInfo commentInfo,
+            NormalizedSnapshotSpanCollection selectedSpans)
+        {
+            var textChanges = ArrayBuilder<TextChange>.GetInstance();
+            var trackingSpans = ArrayBuilder<CommentTrackingSpan>.GetInstance();
+
+            var linesInSelections = selectedSpans.ToDictionary(
+                span => span,
+                span => GetLinesFromSelectedSpan(span).ToImmutableArray());
+
+            Operation operation;
+            // If all the selections are fully commented, uncomment.
+            if (!linesInSelections.Values.Where(lines => SelectionHasUncommentedLines(lines, commentInfo)).Any())
+            {
+                foreach (var selection in linesInSelections)
+                {
+                    UncommentLines(selection.Value, textChanges, trackingSpans, commentInfo);
+                }
+
+                operation = Operation.Uncomment;
+            }
+            else
+            {
+                foreach (var selection in linesInSelections)
+                {
+                    CommentLines(selection.Key, selection.Value, textChanges, trackingSpans, commentInfo);
+                }
+
+                operation = Operation.Comment;
+            }
+
+            return new CommentSelectionResult(textChanges, trackingSpans, operation);
+        }
+
+        private static void UncommentLines(ImmutableArray<ITextSnapshotLine> commentedLines, ArrayBuilder<TextChange> textChanges,
+            ArrayBuilder<CommentTrackingSpan> trackingSpans, CommentSelectionInfo commentInfo)
```

Comment:
i'm a bit surpirsed with this impl.  Don't we already ahve a comment/uncomment feature?  I would have expected the code to be shared between these two user facing features.  that way if 'toggle' chooses to comment, we get the same behavior as the user manually invoking "comment lines" (and vice versa).  Can we share more code here to ensure we're consistent between the features?

Reply:
I think it would be easier to actually refactor the existing feature to be more similar to the new comment behavior.

A few problems that make it hard to re-use existing behavior
1.  a comment action in the existing feature can either use block comments or single line comment depending on the selection.
2.  the new feature needs to handle multicaret selections different from the existing feature because I need to ensure consistency between all selections (where consistency is guaranteed by the existing comment by the command pressed)

1/2 mean that I need to determine upfront all the lines upfront rather than lazily retrieving them when I'm ready to comment, see http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/CommentSelection/CommentUncommentSelectionCommandHandler.cs,47084d65c6aa41c4

I extracted the easy bits out (e.g. the determine indent). 

I could also update the existing comment handler implementation to be more similar to the new implementation, which would allow us to re-use a bit mroe code.

Reply:
I've created an issue for myself to go back and refactor both of the handlers to re-use more of the same code (which will end up refactoring the existing handler mostly)
https://github.com/dotnet/roslyn/issues/34944

I want to get this one into p2, doing it in a followup issue will give me a little more time to properly test the existing comment/uncomment feature when I refactor it :)